### ISSUE TITLE
Problem with brew and dependencies

### ISSUE THREAD

**Issue by AndreaCallia** [dreal/dreal4#81]
Hi all,

when I try to build dReal following the instructions and I type

`./setup/mac/install_prereqs.sh`

then the script tries to run

`brew install dreal --only-dependencies --build-from-source`

But when doing this, it tries to install the `osi` formula, which has been removed recently, and I get the  following error.

```
==> Installing dreal from dreal/dreal
Error: No available formula with the name "dreal-deps/coinor/osi" (dependency of dreal/dreal/dreal)
==> Searching for a previously deleted formula (in the last month)...
Warning: dreal-deps/coinor is shallow clone. To get complete history run:
  git -C "$(brew --repo dreal-deps/coinor)" fetch --unshallow

dreal-deps/coinor/osi was deleted from dreal-deps/coinor in commit df4f51a:
  Remove unused osi.rb

To show the formula before removal run:
  git -C "$(brew --repo dreal-deps/coinor)" show df4f51a^:osi.rb

If you still use this formula consider creating your own tap:
  https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap

```

**Comment by soonho-tri** [comment]
Could you check if you can reproduce the problem after running `brew update`? If so, please post the output from `brew config` here.

**Comment by AndreaCallia** [comment]
The problem persists also after running `brew update`.

This is the output of `Brew config`:

```
HOMEBREW_VERSION: 1.6.0
ORIGIN: https://github.com/Homebrew/brew
HEAD: e3cd9c972f55e9953830f98b4b64eb4ba10727b1
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 376605b56393496731066ad842c36b564cc87299
Core tap last commit: 53 minutes ago
HOMEBREW_PREFIX: /usr/local
CPU: quad-core 64-bit kabylake
Homebrew Ruby: 2.3.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3/bin/ruby
Clang: 9.0 build 900
Git: 2.13.6 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 9.0.1, 1.8.0_162
macOS: 10.12.6-x86_64
CLT: 9.2.0.0.1.1510905681
Xcode: 9.1
XQuartz: 2.7.11 => /opt/X11
```

**Comment by soonho-tri** [comment]
Does the following help?
```bash
brew untap dreal/dreal && brew tap dreal/dreal
brew untap dreal-deps/coinor && brew tap dreal-deps/coinor
brew install dreal --only-dependencies --build-from-source
```

**Comment by soonho-tri** [comment]
Hmm.. I'm wondering why it keeps looking for `osi`. Could you first do `brew cleanup` and remove old version of `clp` if any?

Then,
```bash
cd /usr/local/Homebrew/Library/Taps/dreal-deps/homebrew-coinor
ls -1 *.rb        # Please check if you have osi.rb
grep "osi" *.rb    # This should give nothing
```

**Comment by AndreaCallia** [comment]
Still the same error unfortunately

**Comment by soonho-tri** [comment]
What do you have when you run
```
cd /usr/local/Homebrew/Library/Taps
grep -R "coinor/osi" *
```

**Comment by AndreaCallia** [comment]
If I do this I get an empty output

**Comment by soonho-tri** [comment]
and I guess the following only gives you a single line with `libosip`?

```
cd /usr/local/Homebrew/Library/Taps
grep -R "osi" * | grep "depends_on"
```

**Comment by AndreaCallia** [comment]
Yes, I get:

`homebrew/homebrew-core/Formula/libexosip.rb:  depends_on "libosip"`

**Comment by soonho-tri** [comment]
But `brew deps clp` still shows you `osi`?

**Comment by soonho-tri** [comment]
Also what do you have for `brew tap | grep "dreal"`?

**Comment by AndreaCallia** [comment]
No, it shows:

```
coin_data_netlib
coin_data_sample
coinutils
doxygen
gcc
gmp
isl
libmpc
mpfr
readline
```

**Comment by soonho-tri** [comment]
Everything looks normal to me. Can you do `brew doctor`, fix if there are any issues, and see if it helps?

**Comment by AndreaCallia** [comment]
Many thanks for this...

There are no errors produced by `brew doctor`, only some warnings that I think are not related to dreal.

But if you want to double check, these are the warnings produced by `brew doctor` (I have put [...] for redundant or unrelated lines), do you think that any of them may affect the installation of the `dreal` build dependencies?

```
Warning: Putting non-prefixed coreutils in your path can cause gmp builds to fail.

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libopensmt2.dylib
  /usr/local/lib/libz3.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  [...]
  /usr/local/include/opensmt/Alg.h
  /usr/local/include/opensmt/Alloc.h
  [...]
  /usr/local/include/z3++.h
  /usr/local/include/z3.h
  [...]

Warning: Unbrewed .pc files were found in /usr/local/lib/pkgconfig.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

[...]

Warning: Homebrew's sbin was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting the PATH for example like so
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.bash_profile

Warning: Your Xcode (9.1) is outdated.
Please update to Xcode 9.2 (or delete it).
Xcode can be updated from the App Store.
```

**Comment by soonho-tri** [comment]
All look innocuous to me. Could you check if the following
```
git -C /usr/local/Homebrew/Library/Taps/dreal-deps/homebrew-coinor log -n 1 --oneline
```
gives you
```
e33f4fc (HEAD -> master, origin/master, origin/HEAD) travis: Add support for 10.11
```
?

**Comment by AndreaCallia** [comment]
I have found [this](https://stackoverflow.com/questions/27296188/what-exactly-is-a-grafted-commit-in-a-shallow-clone) about the meaning of _grafted_ commits, then it may be related to this part of the `osi`-related error?

```
Warning: dreal-deps/coinor is shallow clone. To get complete history run:
  git -C "$(brew --repo dreal-deps/coinor)" fetch --unshallow
```

I think this warning is quite related with the error about `osi`

**Comment by soonho-tri** [comment]
I'm not sure. You can check it by running `git -C "$(brew --repo dreal-deps/coinor)" fetch --unshallow` and retest things.

It's still unclear to me why `brew` thinks `dreal` depends on `osi`. Can you check the following?
```
brew deps dreal
brew deps ibex@2.6.5
```

Do they have `osi`?

**Comment by AndreaCallia** [comment]
The don't unfortunately...

It's becoming very subtle, I even tried this:
```
( brew deps dreal ; for dep in $(brew deps dreal) ; do brew deps $dep ; done ) | grep -i osi
```
and still `osi` does not appear. At this point I guess I should try to do a recursive check.

**Comment by AndreaCallia** [comment]
The output of this command suggested by `brew` seems to mention `osi`.
Command:
```
git -C "$(brew --repo dreal-deps/coinor)" show df4f51a^:osi.rb | cat
```

Output:
```
class Osi < Formula
  desc "abstract class to a generic linear programming solver, and derived classes for specific solvers"
  homepage "https://projects.coin-or.org/Osi/"
  url "http://www.coin-or.org/download/pkgsource/Osi/Osi-0.107.5.tgz"
  sha256 "c578256853109e33a8ad8f1917ca8850a027d8a0987ef87166d34de1a256a57d"
  head "https://projects.coin-or.org/svn/Osi/trunk"
  revision 1

  bottle do
    root_url 'https://dl.bintray.com/dreal/homebrew-coinor'
    cellar :any
    sha256 "92eecd19cb42d83c3af251c6d59bf501f5f3055cca733a731b037201235cf542" => :el_capitan
    sha256 "f1c979594abae16bafa90fe33f8d9d7d0ad7abe09da19d38c4991fa050aa0577" => :sierra
    sha256 "fa7ff052cb881819cafdc989995c7960c7ebf53cbfd0dd602db97039223c8fee" => :high_sierra
  end

  option "with-glpk", "Build with interface to GLPK and support for reading AMPL/GMPL models"

  glpk_dep = (build.with? "glpk") ? ["with-glpk"] : []
  openblas_dep = (build.with? "openblas") ? ["with-openblas"] : []

  depends_on "homebrew/science/openblas" => :optional
  depends_on "homebrew/science/glpk448" if build.with? "glpk"

  depends_on "coinutils" => (glpk_dep + openblas_dep)
  depends_on "pkg-config" => :build

  def install
    args = ["--disable-debug",
            "--disable-dependency-tracking",
            "--prefix=#{prefix}",
            "--datadir=#{pkgshare}",
            "--includedir=#{include}/osi",
            "--with-sample-datadir=#{Formula["coin_data_sample"].opt_pkgshare}/coin/Data/Sample",
            "--with-netlib-datadir=#{Formula["coin_data_netlib"].opt_pkgshare}/coin/Data/Netlib",
            "--with-dot",
           ]

    if build.with? "glpk"
      args << "--with-glpk-lib=-L#{Formula["glpk448"].opt_lib} -lglpk"
      args << "--with-glpk-incdir=#{Formula["glpk448"].opt_include}"
    end

    system "./configure", *args

    ENV.deparallelize # make fails in parallel.
    system "make"
    system "make", "test"
    system "make", "install"

    inreplace "#{lib}/pkgconfig/osi-unittests.pc", prefix, opt_prefix
    inreplace "#{lib}/pkgconfig/osi.pc", prefix, opt_prefix
  end
end
```

**Comment by AndreaCallia** [comment]
Just to rule out this, I made a very dirty but effective recursive check using this command:

`deps="dreal" ; while [ "$deps" != "" ] ; do new="" ; for dep in $(echo $deps) ; do if [ ! -f deps/$dep.txt ] ; then new="$new $(brew deps $dep | teed deps/$dep.txt)" ; fi ; done ; deps="$new" ; done`

After that, I checked using
`grep -iR osi deps/*`

Again, I got no results, so there are no dependencies (not even indirect) which require the installation of `osi`. This makes very difficult to understand why `brew` is trying to install it.

At this point, could it be something stored in some cache by `brew` ? I know you suggested `brew cleanup` already, but perhaps there are also other things that need to be cleaned up?

**Comment by AndreaCallia** [comment]
I also tried to manually clear the cache using this command:
```
rm -rf `brew --cache`
```
but even after this, still the same error is generated.

**Comment by AndreaCallia** [comment]
Many thanks for this attempt Soonho, but after running your instructions and after running again:
```
brew install dreal --only-dependencies --build-from-source
```
I still get the same error... I am wondering, since the option `only-dependencies` installs "only" the dependencies, should it be a subset of the formulas installed using this option? Then why if I remove this formula does it work?

In other words, if I run:
```
brew install dreal --build-from-source
```
it works correctly. What do you think? Since this one has worked, should I consider the formulas necessary to build `dreal` as installed?

**Comment by soonho-tri** [comment]
I think there is some inconsistencies either in homebrew or in your homebrew installation.

If `brew install dreal --build-from-source` works, I think you're good to go and ready to build dReal4.

BTW, I just found that there is `--include-build` option in `brew deps`. Can you check if your `dreal` formula include `osi` when you run `brew deps dreal --include-build`?

**Comment by AndreaCallia** [comment]
Many thanks for the advice! I will send the issue to the Homebrew people, if it's true that `--only-dependencies` only installs a subset of the formulas, then it should be impossible that it doesn't work, there must be some inconsistency, I guess due to some not properly cleaned up data (Homebrew must read this "osi" name somewhere in some file). Or some mistake in one of the formulas.

I will let send a confirmation that it worked after trying to build everything. Many thanks again for your great support.

**Comment by AndreaCallia** [comment]
You are right! Thank you very much for this suggestion

**Comment by soonho-tri** [comment]
I guess it worked? Great!

Closing this one now.

**Comment by soonho-tri** [comment]
FYI, I think the problem is caused by the fact that 1) I built and uploaded the dreal/ibex/coinor bottles (pre-built homebrew packges) first when `osi` was a dependency of them, and 2) I removed `osi` dependency later. So homebrew got confused by the inconsistency between the current formulas and their bottles.

1) I added `osi` back to `dreal-deps/homebrew-coinor` to avoid any problems for now.
2) I'll rebuild the bottles and then remove `osi`.
